CPS-0010? | Wallet Connectors (#619)
cardano-foundation/CIPs: CPS-0010/README.md

---
CPS: 10
Title: Wallet Connectors
Status: Open
Category: Wallets
Authors:
    - Ryan Williams <ryan.williams@iohk.io>
    - Vladimir Kalnitsky <vladimir@mlabs.city>

> Proposed Solutions: []

Reviewer: For my own understanding, is this where future adopted Web-Wallet bridges like CIP-90 would go when merged?

Author: Yes, I believe this to be the intension.

from CIP-9999:

> A list of CIPs addressing the problem, if any

Reviewer: thanks @Ryun1 then I assume for this same reason that CIP-30 _does not_ go here... due to its being the _problem_ rather than the _solution_ 😅

Author: > CIP-30 does not go here... due to its being the problem rather than the solution 😅

Yup! 😆
and currently I dont think CIP90 or CIP45 could go here either, as they do not fully satisfy the problem(s) described

Reviewer: yes, all these CIPs seem to have a large factor in common 😆

#### 6. Respect the role of wallet
APIs should be written to have a clear scope, understanding the potential strains placed on wallet providers.
APIs, at a base level, should focus on preserving the base [role of the wallet](#clear-role-of-wallet).

### Context

#### Cardano Connector History
[CIP-30](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0030/README.md) is *the* fundamental wallet connector standard for Cardano.
This standard has facilitated the emergence of dApp development on Cardano by defining both a connection standard and an API.
It is, to date, the only wallet connector standard to see wide adoption in the ecosystem.
Using an injected Javascript object for communication between wallets and web-based stacks
Since its' authoring, CIP-30 has seen continued iteration with changes to its API.

##### CIP-30 Alternatives
Since CIP-30's authoring there have only been two other competing standards, in CIP-45 | Decentralized WebRTC dApp-Wallet Communication ([#395](https://github.com/cardano-foundation/CIPs/pull/395)) and CIP-90 | Extendable dApp-Wallet Web Bridge ([#462](https://github.com/cardano-foundation/CIPs/pull/462/)).

> Although neither of of this standards have seen wide adoption.

Reviewer: ```suggestion
Neither of these standards have seen wide adoption.
```
(grammar) Another way to fix the hanging clause (a sentence fragment) is to run the 2 sentences together with a comma before "Although".

Although not a direct competitor, CIP-13 could be seen as an alternative standard, which fits some wallet connection niches.
Where information is being transmitted to a wallet, to perform an action, using a novel URI scheme.

##### CIP-30 Extensions
The CIP-30 extendibility mechanism was introduced as a novel versioning scheme.

- CIP-62
- CIP-95
- CIP-103

> - CIP-104 -- impact of sharing account public key?

Reviewer: We have another extension on the way without a number yet (#623) but I don't think this list here has to be complete for the CPS to be illustrative the way it stands now.

### dApp Developer
Carol is a dApp developer, who wants many users to be able interact with her dApp via web interface.
She wants to be able to support a wide range of wallets and platforms to maximize potential user base.
Whilst she wants the API standards to be expressive so that her dApp is able to offer a range of functionality.

### Wallet provider
Dave has created a wallet, he wants to minimize the cost of running his wallet.

Without clear role of the wallet, and optional API extension scoping Dave's infrastructure may be asked to do many operations incurring cost.

## Goals
<!-- A list of goals and non-goals a project is pursuing, ranked by importance. These goals should help understand the design space for the solution and what the underlying project is ultimately trying to achieve.

Goals may also contain requirements for the project. For example, they may include anything from a deadline to a budget (in terms of complexity or time) to security concerns.

Finally, goals may also serve as evaluation metrics to assess how good a proposed solution is. -->
### Clear Role of Wallet

> Whilst most wallet softwares generally offer a users a range of features, at their core all wallets are concerned with the management of cryptographic operations.

Reviewer: ```suggestion
Whilst most software wallets generally offer users a range of features, at their core all wallets are concerned with the management of cryptographic operations.
```

> The majority of wallets build on this by using a chain indexer, so information related to the wallets crypto credentials can be gathered and presented to users.
> Any operation beyond core cryptographic functions should be considered to be optionally extra by wallets.
> 
> By only expecting wallet to support cryptographic operations we define a universal rule for all wallet connectors.

Reviewer: ```suggestion
The majority of wallets build on this by using a chain indexer, so information related to wallets' crypto credentials can be gathered and presented to users.
Any operation beyond core cryptographic functions should be considered optional by wallets.

By only expecting wallets to support cryptographic operations we define a universal rule for all wallet connectors.
```
("optionally extra" is redundant)

- **Two sources of truth problem** - on Cardano, every running node has its own opinion on the set of currently unspent transaction outputs. Only [eventual consistency](https://en.wikipedia.org/wiki/Eventual_consistency) is guaranteed. Any dApp that interacts with an own-data wallet has to deal with the inconsistency between local `cardano-node`-based query layer and light wallet query layer, especially when dApp workflow involves sending multiple interactions with the wallet in quick succession. For example, a wallet may refuse to sign a transaction containing an UTxO it does not (yet) know about.

Even if we enforce clear separation of concerns in the standards, e.g. by stating that sending a transaction bypassing the wallet is not allowed, the problem of two sources of truth will remain.

##### Full-data wallets

> Full-data wallets allow to query blockchain data outside of user's scope (i.e. anything not covered by own-data wallets).

Reviewer: ```suggestion
Full-data wallets allow querying blockchain data outside of user's scope (i.e. anything not covered by own-data wallets).
```
(grammar: unlike other languages, in English the verb tense here would be participle rather than infinitive)

![diagram showing possible wallet and dApp architecture for full-data wallets](./full-data-wallet.drawio.png)

Depending on dApp needs, full-data wallets open a way to implement fully-functional dApps that use non-local blockchain data without the need for the developer to maintain dApp backend infrastructure. As a result, more decentralized and censorship-resistant architectures become possible: a dApp can be distributed as a set of files, and deployed to IPFS or static hosting websites.

Full-data wallets enable "single source of truth" architecture: no need to work around data inconsistency between two query layers.

> They are more expensive to operate due to higher query layer requirements, but the risks could be lowered by letting query layer providers and wallet developers be separate entities. That would also allow the end users to choose query layer providers they trust more rather than being forced to trust wallet's infrastructure as well as code, and, optionally, use their own query layer deployments, thus alleviating the need to trust any external wallet backend while enjoying the usual interface.

Reviewer: ```suggestion
They are more expensive to operate due to higher query layer requirements, but the risks could be lowered by letting query layer providers and wallet developers be separate entities. That would also allow the end users to choose query layer providers they trust more, rather than being forced to trust wallet's infrastructure as well as code, and, optionally, use their own query layer deployments, thus alleviating the need to trust any external wallet backend while enjoying the usual interface.
```
(grammar, and also resolves an ambiguity in the reading) matches the comma at the end of this phrase.

Although not a direct competitor, [CIP-13 | Cardano URI Scheme](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0013/README.md) could be seen as an alternative standard, fitting some wallet-client niches.
This URI scheme allows a sort-of one way connection, where information can be passed to a wallet to trigger a function.

#### Historical Issues
CIP-30 is likely the most iterated upon CIP, post merger into the CIPs repository.
These changes have been attempting to remedy issues with the initial proposal.

##### Versioning
Despite the CIP-30 API including a [versioning mechanism](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0030/README.md#cardanowalletnameapiversion-string) it was not utilized.
Alterations to the API were accepted without incrementing the version number.

These changes resulted in diverging implementations and thus broken compatibility.
At connection time there would be no way to tell which version of CIP-30 the wallet had implemented.
Often resulting in errors during dApp-wallet communication, degrading usability.

##### Unclear Responsibilities
CIP-30 does not provide rationale to what should be the concern of the wallet and clients.

> The knock-on effects of this are XXX

Reviewer 2: ```suggestion
The knock-on effect of this is uncertainty regarding the direction future CIP-30 development should follow.
```

##### Language Dependent
CIP-30's connection standard and API is defined using Typescript.
Whilst this is convenient for Javascript-based clients this is not convenient for other stacks. 

Tying the wallet connector to a single language is naturally limiting and will cause friction for potential adopters.

##### Limited to Web-based Stacks
The CIP-30 connection standard is based upon injecting code into shared web windows.
This is convenient for browser extension wallets and web-based client stacks, but is awkward for all other implementors.

One consequence of this is that mobile wallets are thus required to reimplement web browsers in their applications, which is wasted effort.
A similar novel solution would be required for the current full node wallets to be able to utilize CIP-30 compatible dApps.

> This is not not great as non-web-based wallets (and their users) are unable to utilize the benefits of Cardano's web 3.

Reviewer 2: ```suggestion
This is not great as non-web-based wallets (and their users) are unable to utilize the benefits of Cardano's web3.
```

##### Undefined Behavior

> The overall the descriptions of CIP-30 behavior are brief leading to many potential cases of undefined behavior.

Reviewer 2: ```suggestion
The overall descriptions of CIP-30 behavior are brief, leading to many potential cases of undefined behavior.
```

For example; should the results of `.getUtxos()` include UTxOs the wallet wishes to reserve/not spend? or include UTxOs the wallet knows are in the process of being spent? how should these reserved/pending UTxOs factor into the result of `.getBalance()`?
Such behavior is unclear and thus different wallets will implement behavior differently, leading to inconsistent experiences for dApps and users.

##### Combined Standard
CIP-30 defined both a connection standard and an API which must be used on every connection.
This is limiting as it forces all connections using the CIP-30 connect to always use the same endpoints.
This adds unneeded complexity, making iterating on connection or API more difficult.

##### No event listener
The CIP-30 API and connector is purely based on synchronous and asynchronous calls made by the client dApp.

> This prevents useful advantages of event-based design, such as dApps subscribing wallets sharing state update events.

Reviewer 2: ```suggestion
This prevents useful advantages of event-based design, such as dApps subscribing to state update events emitted by wallets.
```

Goals may also contain requirements for the project. For example, they may include anything from a deadline to a budget (in terms of complexity or time) to security concerns.

Finally, goals may also serve as evaluation metrics to assess how good a proposed solution is. -->
These goals we outline for future wallet connection and API standards are based upon solving our [Core Concerns](#core-concerns).

**Wallet connectors should:**

### 1. Prioritize Security
Potential security concerns of wallet connector designs should be well understood and discussed in accompanying CIPs.
Authors should seek the collective knowledge of the CIP process to obtain wide range of perspectives.

The impact of any data leaving the wallet should be discussed, especially for the cases of data that is not observable via the chain, such as root public keys.

Although security tolerances are at an implementors discretion, the potential negative impact on the ecosystem should be taken into account by authors.

> ### 2. Interoperable, Optional and Extensible APIs

Reviewer 2: ```suggestion
### 2. Provide Interoperable, Optional and Extensible APIs
```

Should be aligned with "Wallet connectors should" above.

Future API standards should be connection standard agnostic.
Meaning API CIP-1234 should work if connection has been initiated via CIP-30 connection or CIP-45, or CIP-4321.

API standards should exist within an extensible framework, whereby specific connections can be permissioned and specialized.
APIs should be developed via the CIP process, whereby CIPs will each define their own set of endpoints.

Implementing the API support should be fully at the discretion of the wallet providers.
Although at connection time dApps should be free to request as many APIs as they wish, they should not rely on the implementation of optional APIs beyond APIs implemented in [No-data Wallets](#no-data-wallets).

> ### 3. Version

Reviewer 2: ```suggestion
### 3. Support Versioning
```

### Can a universal connector be pursued?
Every wallet connector that simply adds functionality on top of another standard can be considered a valid implementation of the base standard.
If the set of standards is designed in a way that allows for extensions without conflicts, then a superset of all standards can be considered a universal connector. How to make it possible to clearly define a "universal connector" and whether it is useful at all remains to be seen.

### Can a universal API be pursued?
Platform differences are a concern: by using specifications that do not rely on assumptions about data transfer methods, software environment or programming language, it is possible to create a standard that is truly reusable across platforms.

There are many types of dApp architectures enabled by different wallet types, but there is no requirement to use the APIs provided by wallets, so dApps that use simpler standards should still remain functional when used with their extensions.

Does that open a way to define a universal API?

> - what language

Reviewer 2: ```suggestion
```

### How interoperable can standards be with those of other ecosystems?
Cardano is sufficiently different from most of the blockchains that aim to unify their wallet standards.

What options do we have for integration with Cardano sidechains, rollups, hardforks, private testnets, etc.?

### How can we effectively police API scope?

> - how can we prevent spams of overlapping APIs

Reviewer 2: ```suggestion
How can we prevent duplication of functionality and make sure different APIs do not overlap?
```
